File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque

def solution(maps):
    def bfs(x, y):
        q = deque()
        q.append([x, y])
        visited[x][y] = True
        days = 0
        while q:
            cx, cy = q.popleft()
            days += int(maps[cx][cy])
            for d in direct:
                nx, ny = cx + d[0], cy + d[1]
                if 0 <= nx < N and 0 <= ny < M and not visited[nx][ny] and maps[nx][ny] != "X":
                    q.append([nx, ny])
                    visited[nx][ny] = True
        return days

    answer = []
    N = len(maps)
    M = len(maps[0])
    visited = [[False] * M for _ in range(N)]
    
    direct = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    for i in range(N):
        for j in range(M):
            if maps[i][j] != "X" and not visited[i][j]:
                answer.append(bfs(i, j))
    
    if answer:
        return sorted(answer)
    else:
        return [-1]

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    const answer = [];
    const [X, Y] = [maps.length, maps[0].length];
    const visited = Array.from(Array(X), () => Array(Y).fill(false));
    const direct = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    
    const bfs = (a, b) => {
        let cnt = 0
        const q = [[a, b]];
        
        cnt += parseInt(maps[a][b]);
        visited[a][b] = true
        
        while (q.length > 0) {
            const [x, y] = q.shift()
            
            for (let d of direct) {
                const nx = x + d[0]
                const ny = y + d[1]
                if (nx >= 0 && nx < X && ny >= 0 && ny < Y && !visited[nx][ny] && maps[nx][ny] !== "X") {
                    visited[nx][ny] = true
                    cnt += parseInt(maps[nx][ny])
                    q.push([nx, ny])
                }
            }
        }
        answer.push(cnt);
    }
    
    for (let i=0; i < X; i++) {
        for (let j=0; j < Y; j++) {
            if (!visited[i][j] && maps[i][j] !== "X") {
                bfs(i, j)
            }
        }
    }
    
    if (answer.length === 0) {
        return [-1]
    } else {
        return answer.sort((a, b) => a - b);   
    }
}